RFC: Light DOM Support (#44)
salesforce/lwc-rfcs: text/0000-light-dom.md

- [MS Fast Element](https://fast.design/docs/fast-element/working-with-shadow-dom#shadow-dom-configuration)

## Detailed design

### Selecting Light DOM vs Shadow DOM

In raw JS, developer has the choice to use or not use shadow DOM for their custom element. For example,

```js
// uses Shadow DOM
class MyShadowElement extends HTMLElement {
  connectedCallback() {
    this.attachShadow({ mode: "open" });
    this.shadowRoot.innerHTML = "<p>My custom element that uses Shadow DOM";
  }
}

// doesn't use Shadow DOM
class MyLightElement extends HTMLElement {
  connectedCallback() {
    this.innerHTML = "<p>My element that does not use Shadow DOM";
  }
}
```

In LWC as well, the developer will retain the choice: they can choose to inherit from `LightningElement` to select Shadow DOM (which already exists) or inherit from a new class, `MacroElement`, to opt into Light DOM.

```js
import { MacroElement } from "lwc";

export default class MyLightComponent extends MacroElement {
  renderedCallback() {
    this.querySelector("child-element"); // note the absence of this.template
  }
}
```

> `MacroElement` is a new class that will be exported from the `lwc` module. Its API and functionality remain identical to `LightningElement` with the primary difference of not using Shadow DOM.

Reviewer: I'm a bit concerned about exposing light DOM via `MacroElement`. I can break my concerns into two groups: 1) deviation from other WC frameworks, and 2) potential class proliferation.

## Deviation from other WC frameworks

Here is how other frameworks handle light DOM:

[Stencil](https://stenciljs.com/docs/host-element#shadow-dom)

```js
@Component({
  tag: 'my-component',
  shadow: false // or omitted
})
class MyComponent {}
```

[Fast](https://www.fast.design/docs/fast-element/working-with-shadow-dom/#shadow-dom-configuration)

```js
@customElement({
  name: 'my-component',
  template,
  shadowOptions: null // or `{ mode: 'closed' }` for closed shadow
})
class MyComponent extends FASTElement {}
```

[Lit](https://lit-element.polymer-project.org/api/classes/_lit_element_.litelement.html#createrenderroot)

```js
class MyComponent extends LitElement {
  createRenderRoot() {
    return this // or return `this.attachShadow(...)` here
  }
}
```

So we have 2 frameworks opting for decorators, and another opting for an overridable method.  LWC would be the only WC library using separate classes (unless I missed one).

## Potential class proliferation

In the future, we may want to support closed shadow roots. Would that be another class?

```js
import { ClosedShadowElement } from 'lwc'
```

There is also a proposal for "open stylable roots" (https://github.com/WICG/webcomponents/issues/909). Would this be another one?

```js
import { OpenStylableElement } from 'lwc'
```

In that proposal, there is even some talk of `attachShadow()` eventually looking like this:

```js
this.attachShadow({
  mode: 'open',
  styles: 'closed'
})
```

If the options bag for `attachShadow` continues to grow like this, then we could end up with a combinatorial explosion of classes to handle every possible case.

My personal preference would be for decorators or [mixins](https://justinfagnani.com/2015/12/21/real-mixins-with-javascript-classes/), but mostly I'm just concerned about choosing classes here.

Reviewer 2: I think the original reason we chose a different class was to emphasize that these are different _kinds_ of components because the programming model is different. I'm not sure if we still feel that is the case. 

I'm curious how the [createElement](https://github.com/salesforce/lwc/blob/ff6c28bc001b55cdee2a6f9442ce5e95a4f0e96a/packages/%40lwc/engine-dom/src/apis/create-element.ts#L89) API will change which currently takes the mode.

Reviewer: @tedconn Ah my mistake, I didn't even know we supported closed shadow roots. 🙂  It seems that `createElement` would be the right place to put the options bag (if it becomes a big bag!).

So it looks like my concerns about class proliferation might not apply, since we only plan on having `MacroElement` for this one case (light DOM). But then why not have light DOM as an option in `createElement`?

Reviewer 2: > But then why not have light DOM as an option in `createElement`

Because we need an API on the component, so component authors can toggle, not on the engine. Something like this would be nicer though, right?
```js
import { LightningElement } from 'lwc';

export default class LightDomElement extends LightningElement {

  static shadowOptions = {
    mode: null
  }

}
```

Reviewer 3: We discussed in the past 2 options, using a static field or using a brand new base class. Until the decorator proposal goes back to stage 3, I don't think it is safe to introduce a brand new decorator.  

The main advantage of introducing a brand new class over static fields is around component inheritance. You can shadow (by inadvertence or intentionally) a base class static field. It means that you can turn a Shadow DOM component into a Light DOM one (and vice-versa) by extending from it. Such override can't be done when inheriting from a different base class.

In the following example, let say that we introduce a new `shadowDOM` static boolean to indicate whether the component rendered in the Light DOM or in the Shadow DOM. It is something we should discourage.

```
import { LigthningElement } from 'lwc';

class ShadowDOMComponent extends LightningElement {
  static shadowDOM = true;
}

class LigthDOMComponent extends ShadowDOMComponent {
  static shadowDOM = false;
}
``` ​

Now that you speak about class proliferation, I am more favorable to adding a signal via a static field.

Reviewer: After discussion with the team, we landed on the static field rather than `MacroElement`, correct? AFAICR, we concluded:

- Any solution we land on (classes, field, method, decorators, etc.) needs to be resolved at runtime. In other words, developers can always do shenanigans such that you need to do a runtime evaluation to figure out whether the element is shadow or light. Since the LWC compiler to know at compile-time, we would just have to teach developers to use patterns that are friendly for compile-time evaluation.
- Decorators are not standard yet
- Classes may result in class proliferation
- Static fields seem to have nice ergonomics
- So we should go with static fields

Is this right? @pmdartus @abdulsattar

### Component features when using Light DOM

Some of the LWC component capabilities are directly inherited from Shadow DOM, or emulated by the synthetic-shadow. Despite the use of Light DOM, we’d like to keep these features available, even if their behavior is slightly adapted to the Light DOM:

- **Slots**

  As we mentioned before, the component composition model in LWC is provided by slots. Light DOM will provide the same mental model for developers building Light DOM components.

  In Light DOM, `<slot>` will denote the place where the slotted component will be attached. The `<slot>` element itself won't be rendered. The slotted content (or the fallback content) will be flattened to the parent element at runtime.

>   Since the `<slot>` element itself isn't rendered, adding attributes or event listeners to the `<slot>` element in the template will throw a compiler error.

Reviewer: Out of Stencil, Fast, and Lit, only Stencil opted to make slots "work" with light DOM.

Fast [says](https://www.fast.design/docs/fast-element/working-with-shadow-dom):

> If you choose to render to the Light DOM, you will not be able to compose the content, use slots, or leverage encapsulated styles.

Lit is [also](https://github.com/Polymer/lit-element/issues/553) [against](https://github.com/WICG/webcomponents/issues/909#issuecomment-743269886) slots in light DOM:

> I don't think `<slot>` makes much sense outside of shadow DOM [...]

Based on discussion [in this thread](https://github.com/WICG/webcomponents/issues/909#issuecomment-743095184), it seems like Stencil's choice may have resulted in significant performance overhead, which is understandable given that a lot of behavior you get "for free" with slots in shadow DOM now has to be reimplemented for light DOM.

Before committing to this, I'd be interested to see some more details about how we plan to handle slots in light DOM:

- How does MutationObserver work?
- How does the timing between slotted versus non-slotted children work?
- Would CSS selectors be able to traverse these boundaries?

I'd also be interested to see some benchmarks, so understand if the issues Stencil ran into were just problems with Stencil's implementation, or if they're inherent to trying to make slots "work" in light DOM.

Reviewer: Another interesting point of comparison: [Aurelia has slots](https://docs.aurelia.io/app-basics/components-revisited#au-slot), but they explicitly mark it as `<au-slot>`, not `<slot>`:

> An obvious question might be "Why not simply 'turn off' shadow DOM, and use the `slot` itself"? We feel that goes to the opposite direction of Aurelia's promise of keeping things as close to native behavior as possible. Moreover, using a different name like `au-slot` makes it clear that the native slot is not used in this case, however still bringing slotting behavior to use.

> - **Scoped Styles**

Reviewer 2: @abdulsattar We need a section to describe how DOM querying will work.

Author: Added.

  Shadow DOM styles are scoped to the enclosing shadow tree. Styles don't leak out to the rest of the page and the page styles don't leak into this component. In Native shadow, it is enforced by the browser whereas in Synthetic shadow it is enforced by adding few attributes to the elements and scoping all CSS rules by those attributes.

  Styles scoping in Light DOM will be done differently. The styles will be scoped not just to the component they belong, but also to its children. These styles won't leak out to the parents (or their children) though.

  E.g.

  ```html
  <x-a>
    <style>
      p {
        color: blue;
      }
    </style>

    <p></p> <!-- will be blue -->
    <p class="red"></p> <!-- will still be blue. red class from child won't leak out here -->

    <x-b>
      <style>
        p.red {
          color: red;
        }
      </style>

      <p></p> <!-- will be blue. Style from x-a will be applied -->
      <p class="red"></p> <!-- will be red -->
      <p class="red"></p> <!-- This paragraph is coming from A and slotted into B. Will also be red. -->
    </x-b>
  </x-a>
  ```

>   Opting out of this scoping is not supported. There's no way for a component author to say a CSS rule shouldn't be scoped to that specific component (and its children). If global scoping is desired, a global stylesheet can be injected manually.

Reviewer: With synthetic shadow DOM, we implemented a system for style scoping similar to this (using attribute selectors). However, with shadow DOM, it was kind of implied that eventually we would migrate to native shadow, so the synthetic system wasn't designed to last forever. With this system, though, we are committing to support a new styling system indefinitely, including whatever implementation details we land on, which will impact compatibility moving forward. For instance:

1. Whether to use attribute selectors / classes / etc. for scoping? This has performance implications (browsers tend to have more optimizations for class than for attribute selectors) as well as specificity implications (e.g. specificity conflicts with global styles). FWIW, Vue uses attribute selectors whereas Svelte uses class selectors.
2. Whether styles are inherited to the children or not. This also seems to me to have performance implications, since I'm not sure how we could implement this, except by concatenating attributes down the descendant tree. (So for instance, would the 5th light DOM grandchild have 5 separate attributes on every DOM node?)
3. Whether styles are inherited to light DOM "slots" or not, and whether something like (for instance) sibling selectors would work across both "slot" and non-"slot" light DOM children

It seems to me that the safest option is to not provide any kind of style scoping for light DOM elements. I.e. "if you want style scoping, use shadow DOM." This avoids making any indefinite commitments toward one particular styling system in light DOM.

Reviewer 2: I think the goal is to land on the most natural ergonomics for the developer, including developers already versed in LWC. 

Styles in LWC are scoped today, but for end users this is not a natural consequence of the shadow DOM encapsulation, rather it just makes sense since the authoring model is per-component. This is why Vue scopes styles, not because it aligns with the Shadow DOM model. For us that is just a bonus.

Not scoping styles would mean that without any extra code on our side, styles would leak into shadow DOM children, which would not match the native shadow DOM behavior when we turn off synthetic.

Reviewer: @tedconn 

> Styles in LWC are scoped today, but for end users this is not a natural consequence of the shadow DOM encapsulation

It is due to `synthetic-shadow`, sure, but the long-term goal is to migrate to native shadow DOM. So then it would be a natural consequence. 🙂 

> This is why Vue scopes styles, not because it aligns with the Shadow DOM model.

I'm not saying scoped styles aren't good for developer ergonomics; I cede that point. 🙂  My point is just that "which scoped styles?" is an important question, especially since we have to indefinitely maintain whatever we choose today.

> Not scoping styles would mean that without any extra code on our side, styles would leak into shadow DOM children

I'm not sure I follow this. Today, in a purely non-LWC, non-synthetic world, I can have a light DOM-using component with a shadow DOM-using component as its child, and the only styles that leak in are [inheritable styles](https://open-wc.org/guides/knowledge/styling/styles-piercing-shadow-dom/) like `font-family` and `color`, which is per the spec. It should work this way both in native shadow DOM and in our synthetic shadow DOM.

Reviewer 2: > It should work this way both in native shadow DOM and in our synthetic shadow DOM.

I may be wrong but my thinking was that If styles aren't scoped, in synthetic shadow DOM, those styles will leak in, because there's no other mechanism to prevent that.

Reviewer: I wasn't sure, so I just ran a quick check. It looks like if you add the style `body { font-family: monospace; }` to a Salesforce page, it will indeed leak in to synthetic shadow DOM components. (This is the same as what would happen with a light DOM-using component wrapping a shadow DOM-using component.) So it appears we are following the spec correctly on that.

Reviewer 3: >  those styles will leak in, because there's no other mechanism to prevent that.

@tedconn You can prevent inherited styles to be applied by resetting all the properties on the host element. It works in both native shadow and synthetic shadow. 

```css
:host {
  all: initial;
}
```

----

We discussed CSS scoping at length and we reached the conclusion that it something we would like to preserve with Light DOM for developer ergonomic reasons, as @tedconn mentioned. 

We considered the following approach to style scoping:

1. **component content scoping:** Styles are applied to the content the component renders and aren't applied to children components. This approach is similar to the way styles are scoped today in synthetic shadow. It also behaves the same as Vue, Svelte, and Angular. An HTML attribute is added to all the CSS selectors, the same attribute is also added to all the elements rendered by the component.

2. **component subtree scoping:** Styles are applied to the content the component renders and also to the content rendered in the children's components. This is a similar approach to Aura styling.  All the CSS selectors are prefixed with a CSS attribute selector and the same attribute is applied to the root element to scope the subtree.

```css
/* Input */
.foo .bar {}

/* Ouput: component content scoping */
.foo[data-scope] .bar[data-scope] {}

/* Output: component subtree scoping */
[data-scope] .foo .bar {}
```

We decided to go with the **component subtree scoping** approach for the following reasons:
- This approach map with the way `querySelector` works in the light DOM. It applies to the entire subtree and not only the current element.
- One of the main reason for using light DOM is to let a component author override it children components style. This is possible with the subtree scoping approach but not with the content scoping one. Overriding children component styles with the subtree scoping approach would require the introduction of a non-standard CSS selector to escape the scoping (`>>>` in [Vue](https://vue-loader.vuejs.org/guide/scoped-css.html#deep-selectors) and [Angular](https://angular.io/guide/component-styles#deprecated-deep--and-ng-deep) or `:global` in [Svelte](https://svelte.dev/docs#style)).

The main downside for me of the  **component subtree scoping** approach is that you can run into a style clashing issue where the parent component can unintentionally override children's component styles because the parent selector has a higher specificity.

Reviewer: Thanks, that makes sense. Two comments:

First, I'm not sure that preferring Aura's styling model to Vue/Svelte's styling model is the best choice from the perspective of developer ergonomics. Aura is well-known inside Salesforce, but not so much outside. (Also if I'm not mistaken, most React CSS-in-JS libraries follow something more like the Vue/Svelte model.) Personally I would find it a bit surprising that parent light components can affect child light components' styles.

Second, for this CSS:

```css
[data-scope] .foo .bar {}
```

I find this to be a bit scary from a performance perspective. Imagine something like:

```css
/* Input */
div {}

/* Output */
[data-scope] div {}
```

Now you're potentially the telling the browser to check the ancestor chain for every `div` to see if its attributes match. For class selectors, there is already a Bloom filter to handle this, but I'm not sure if browsers have gotten around to optimizing attribute selectors in the same way. I may have to whip up a benchmark to test this. 🙂

Reviewer 3: > Personally I would find it a bit surprising that parent light components can affect child light components' styles.

This is good to know. We can survey developers internally to see which approach they find the most intuitive.

> Now you're potentially the telling the browser to check the ancestor chain for every div to see if its attributes match. 

I don't think we can use classes because component developers can set classes via the template (from outside) or in the component via `this.classList` (from within). We would need to special-case this scoping class and ensure that it is always present even when the author manipulates the component classes. I am not sure if we can do this in an efficient way.

```js
import { MacroElement } from "lwc";

export default class MyLightComponent extends MacroElement {
  renderedCallback() {
    this.querySelector("child-element"); // note the absence of this.template
  }
}
```

`MacroElement` is a new class that will be exported from the `lwc` module. Its API and functionality remain identical to `LightningElement` with the primary difference of not using Shadow DOM.

### Component features when using Light DOM

Some of the LWC component capabilities are directly inherited from Shadow DOM, or emulated by the synthetic-shadow. Despite the use of Light DOM, we’d like to keep these features available, even if their behavior is slightly adapted to the Light DOM:

#### Slots

As we mentioned before, the component composition model in LWC is provided by slots. Light DOM will provide the same mental model for developers building Light DOM components.

In Light DOM, `<slot>` will denote the place where the slotted component will be attached. The `<slot>` element itself won't be rendered. The slotted content (or the fallback content) will be flattened to the parent element at runtime.

> Since the `<slot>` element itself isn't rendered, adding attributes or event listeners to the `<slot>` element in the template will throw a compiler error.

Reviewer: I may be misremembering, but I thought light DOM wouldn't have slots? Should this section be removed? (Or replaced with a stub saying "If you want slots, use shadow DOM"? 😃 )

Reviewer: Based on discussion, it looks like this needs to be updated with the new `macro:slot` proposal. Other questions that are worth answering in this section:

What is the timing model of slots? For instance:

```html
<foo-component>
<macro:slot></macro:slot>
<bar-component>
```

In what order do the connectedCallbacks for `foo-component`, elements within the slot, and `bar-component` get called?

Related to that question: will light DOM slots be [lazy or eager](https://dev.to/richharris/why-i-don-t-use-web-components-2cia)? My preference is for lazy, for the reasons outlined in the linked post, e.g.:

```html
<template if:true={foo}>
  <macro:slot></macro:slot>  
</template>
```

In the above case, rendering the slot is wasted effort if `foo` evaluates to `false`. Lazy slots would allow us to skip that wasted work.

Also: how will slots impact SSR? I assume SSR will just render the slotted content directly, i.e. `<macro:slot>` is never actually represented in DOM that the browser sees? But then I assume there will need to be some HTML comments so that hydration knows where the slots are. This can probably be handled in SSR RFCs though.

Reviewer: Another question about timing in slots. Based on my reading of https://github.com/WICG/webcomponents/issues/909#issuecomment-743269886, there is an issue that arises with e.g.:

```html
<!-- child-component.html -->
<div>hello</div>
<macro:slot></macro:slot>
<macro:slot name="title"></macro:slot>
<div>world</div>
```

```html
<!-- parent-component.html -->
<child-component>
  <h1 macro:slot="title"></h1>
  <div>yolo</div>
</child-component>
```

There are two options here: 1. the user's content renders first, or 2) the `child-component`'s content renders first.

In either case, the two content sources could overwrite each other, e.g. by code in the `connectedCallback` or `renderedCallback` appending elements to `this`, or by reading `element.nextSibling`, etc. Because there is no shadow boundary, the two DOM sources can observe each other.

Also, note that the `title` slot and the bare slot have a different ordering in the user's code versus the `child-component`'s code. So in the `renderedCallback` for each, what element order is observed?

In that comment, the author suggests that the only way to make this work is to render the user's content into its own little isolated DOM and then "project" it into the other part of the DOM. But clearly this would involve MutationObserver or some other heavyweight mechanism to mirror the two. It definitely seems complicated.

Prior art worth investigating: [Svelte slots](https://svelte.dev/docs#slot) and [Vue slots](https://vuejs.org/v2/guide/components-slots.html).

Reviewer: FWIW I tried to reproduce the performance problems with light-DOM slots described in [this post](https://medium.com/8451/make-your-stenciljs-web-components-faster-by-using-shadow-dom-d010a9f0cdda), but I can't find a repro.

Stencil's light DOM slot implementation seems to not suffer from significant costs from "moving" the DOM nodes around, so maybe there isn't really a concern here.

Reviewer: Updated RFC with some clarifications on how slots will work.

#### Styles

> Until now styles used in LWC components where scoped to the component thanks to shadow DOM (or synthetic shadow DOM) style scoping. In the light DOM, component styles are naturally leaks out of the component, LWC doesn't do any style scoping out of the box. Developers are in charge of making sure to write specific enough selector to target the component content.

Reviewer: ```suggestion
Until now styles used in LWC components were scoped to the component thanks to shadow DOM (or synthetic shadow DOM) style scoping. In the light DOM, component styles naturally leak out of the component; LWC doesn't do any style scoping out of the box. Developers are in charge of making sure to write specific enough selectors to target the intended element or pseudo-element.
```

> To support the cases where a shadow DOM element composes a light element, light DOM styles are required to be injected to the closest root node. For a given light DOM elements if all the ancestor components are also a light DOM components, the component style sheet will be injected in the document `<head>`. Otherwise if any of the ancestors is a shadow DOM components the style has to be injected in the closest shadow root. Upon insertion of a light DOM element does the following steps:

Reviewer: ```suggestion
To support the cases where a shadow DOM element composes a light element, light DOM styles are required to be injected to the closest root node. For a given light DOM elements if all the ancestor components are also a light DOM components, the component style sheet will be injected in the document `<head>`. Otherwise if any of the ancestors is a shadow DOM component, the style has to be injected in the closest shadow root. Upon insertion of a light DOM element, LWC does the following steps:
```

- look up for the closest root node (`Node.prototype.getRootNode()`)
- insert the stylesheet if not already present:
  - if the root node is the HTML document, the style sheet is inserted as a `<style>` element in the `<head>`.

>   - if the root node is a shadow root, the stylesheet is inserted as a `<style>` element as the first shadow root child.

Reviewer: This would be a good opportunity to align with [constructable stylesheets](https://developers.google.com/web/updates/2019/02/constructable-stylesheets), which is already available in Chrome and fairly easy to polyfill/workaround in other browsers.

The main benefit of this API is that, if some custom element is repeated several times in the document, then this is an explicit signal to the browser that it can optimize that stylesheet's use across multiple elements. Otherwise if you have 1,000 shadow DOM elements each with an identical light DOM element inside, we would have to create 1,000 identical `<style>`s and inject them into the `shadowRoot`. (Presumably browsers could optimize this based on a checksum of the `<style>` content or something, but for whatever reason the Chrome devs seem to prefer constructable stylesheets as a more explicit signal.)

I would propose using constructable stylesheets in Chrome, and the above steps for other browsers. In the case of Chrome, it's just a matter of doing the exact same thing (targeting either the `head` or the nearest `shadowRoot`) but instead of inserting a `<style>` element, we append to [`documentOrShadowRoot.adoptedStyleSheets`](https://wicg.github.io/construct-stylesheets/#proposed-solution).

(N.B. this would require disallowing `@import`, but I assume we already do this? If not, we would have to.)

Off-topic: we could already switch to using this today for synthetic shadow DOM, AFAICT. /cc @ravijayaramappa

Reviewer: @pmdartus Do these `@import` statements actually end up in the final CSS, or is it just an authoring concern? If it ends up in the final CSS, then using constructable stylesheets gets much more complicated, because we would need to use the async API rather than the sync API.

Reviewer 3: No, the imports are stripped out. Those imports are resolved by the application container. This is why the [`TemplateStylesheetFactories`](https://github.com/salesforce/lwc/blob/15ff9bb98ec324dc68d40c96e6d37de69d9f1145/packages/%40lwc/engine-core/src/framework/stylesheet.ts#L25-L30) is recursive as it might contain other stylesheets.

Reviewer: Constructable stylesheets may be a perf optimization we can do later. Resolving for now.

> It is important to notice that the order in which light DOM components are rendered impact the order in which stylesheets are injected in the root node and directly influences CSS rule specificity.

Reviewer: I wonder if it would be better to leave this specificity detail as undefined behavior? Otherwise in the future, we may wish to do some fancy pre-compilation that eagerly injects the stylesheets into the DOM, in which case the order isn't guaranteed… Or maybe I'm overthinking it, and this will never be a concern. 😄

> **Note:** Different approaches layer style scoping on top has been discussed while designing Light DOM, like introducing a new file extension for automatic style scoping `.scoped.css` or using a `<style scoped>` element in the template.

Reviewer: ```suggestion
**Note:** Different approaches to layer style scoping on top has been discussed while designing Light DOM, like introducing a new file extension for automatic style scoping (`.scoped.css`) or using a `<style scoped>` element in the template. This can be addressed in a future RFC.
```

#### `this.template`

In `LightningElement`, `this.template` returns the shadow-root. It will return `null` in `MacroElement`.

### Querying

> `MacroElement` (and `LightningElement`) forward several DOM querying/manipulation methods like `querySelector`, `dispatchEvent` etc. to the host element. Full list of methods forwarded [here](https://github.com/salesforce/lwc/blob/master/packages/@lwc/engine-core/src/framework/base-lightning-element.ts#L242).

Reviewer: ```suggestion
`MacroElement` (and `LightningElement`) forward several DOM querying/manipulation methods like `querySelector`, `dispatchEvent` etc. to the host element. Full list of methods forwarded [here](https://github.com/salesforce/lwc/blob/15ff9bb98ec324dc68d40c96e6d37de69d9f1145/packages/%40lwc/engine-core/src/framework/base-lightning-element.ts#L134-L160).
```

I believe that by not using a permalink, the code changed and the link is no longer valid. Is this the code that was supposed to be referenced?

This will allow component author to just use `this.querySelector` instead of using `this.template.querySelector` in case of `LightningElement`.

> However, there's no way to get a direct reference to the host element, like `this.template.host` that's available in `LightningElement`.

Reviewer: I'm confused by this section. In the context of a light DOM component, `this` refers to the host element, correct? (At least, it does with vanilla web components.)

I'm also not sure what the example below is supposed to be illustrating? Is the example below about event retargeting when a light DOM component has a shadow DOM component as its child? What is the example communicating? (Maybe some code comments would help. 🙂 )

E.g.

```js
/* <template>
 *   <x-child></x-child>
 * </template>
 */
export default class ParentElement extends MacroElement {
  renderedCallback() {
    this.querySelector("x-child").addEventListener("custom", (e) => {
      console.log("Called with " + JSON.stringify(e.detail));
    });
  }
}

/**
 * <template>
 *   <button>Click Me</button>
 * </template>
 */
export default class ChildElement extends MacroElement {
  renderedCallback() {
    this.querySelector("button").addEventListener("click", this.handleClick);
  }

>   handleClick() {
>     this.dispatchEvent(new CustomEvent("custom", { detail: { macro: true } }));
>   }

Reviewer: ```suggestion
  handleClick = () => {
    this.dispatchEvent(new CustomEvent("custom", { detail: { macro: true } }));
  }
```

Without this, the callback doesn't actually get a `this` that refers to the component.

}
```

### Security (WIP)

> - In some applications, light-dom components may not be allowed... it’s up to the app context

Reviewer: Do we have an example of how light DOM components can be disabled? Will there be a compiler flag to accomplish this?

This feature should be exposed and explained to the component library developers as they might change how they develop their components internally.

## How we teach this

Shadow DOM and Light DOM are already names accepted by the industry, see: [Terminology: light DOM vs. shadow DOM](https://developers.google.com/web/fundamentals/web-components/shadowdom?hl=en).
We need to provide the proper documentation to educate the LWC developers:

- What are the differences between Shadow DOM and Light DOM
- We need a guide on when to use one or the other

## Open questions

**How to deal with different namespaced packages claiming the same custom element name?** 

As of today on the Salesforce platform, all the LWC components are authored under the `c` namespace. This namespace dictates the prefix for of the custom elements rendered in the DOM. When delivered via a namespaced package, the components are referenced using their actual namespace (eg. `lightning`, `my_ns`, etc.). However internal component references in the package is done using the `c` namespace. 

> Because of this if 2 manage package publish a `button` component, the 2 components will be rendered as `c-button` in DOM. This is problematic because a custom element name can only be bound to a single element constructor. Even if this is not a new issue, the [scoped custom element registry proposal](https://github.com/WICG/webcomponents/issues/716) would solve this issue for shadow DOM components. This is still an issue for light DOM component as scoped custom elements registry can only be attached to a shadow root.  

Reviewer: Is this a LEX-specific issue, or can it be abstracted to apps not using LEX? Or is it more in the wheelhouse of Locker?

Reviewer 3: This is a Salesforce platform-specific issue. We would run into this issue in LEX and in Commerce Cloud as well. Even if the Locker team have an opinion, it certainly something the LWC team has to drive.

In raw JS, developer has the choice to use or not use shadow DOM for their custom element. For example,

```js
// uses Shadow DOM
class MyShadowElement extends HTMLElement {
  connectedCallback() {
    this.attachShadow({ mode: "open" });
    this.shadowRoot.innerHTML = "<p>My custom element that uses Shadow DOM";
  }
}

// doesn't use Shadow DOM
class MyLightElement extends HTMLElement {
  connectedCallback() {
    this.innerHTML = "<p>My element that does not use Shadow DOM";
  }
}
```

In LWC as well, the developer will retain the choice: they can choose to inherit from `LightningElement` to select Shadow DOM (which already exists) or inherit from a new class, `MacroElement`, to opt into Light DOM.

```js
import { MacroElement } from "lwc";

export default class MyLightComponent extends MacroElement {
  renderedCallback() {
    this.querySelector("child-element"); // note the absence of this.template
  }
}
```

`MacroElement` is a new class that will be exported from the `lwc` module. Its API and functionality remain identical to `LightningElement` with the primary difference of not using Shadow DOM.

### Component features when using Light DOM

> Some of the LWC component capabilities are directly inherited from Shadow DOM, or emulated by the synthetic-shadow. Despite the use of Light DOM, we’d like to keep these features available, even if their behavior is slightly adapted to the Light DOM:

Reviewer 4: Can you describe how events and handlers will work on Light DOM? For example: when an event is dispatched within a slot, event retargeting, etc.

Author: I'll add more detail on that.

However, there won't be a slot, so there won't be any events getting dispatched from the slot. Also, since there's no Shadow DOM, there won't be event retargeting. Not having event retargeting is one of the goals because some third party libraries (like Google Tag Manager) rely on `event.target` to track events. Event retargeting was getting in the way.

Reviewer: Added some notes on `slotchange`, and also we have this line:

> The `<slot>` element itself won't be rendered.

...which I think answers these questions. So I believe we can resolve this conversation.

  
  This will allow component author to just use `this.querySelector` instead of using `this.template.querySelector` in case of `LightningElement`.
  However, there's no way to get a direct reference to the host element, like `this.template.host` that's available in `LightningElement`.

  E.g.
  ```js
  /* <template>
   *   <x-child></x-child>
   * </template>
   */
  export default class ParentElement extends MacroElement {
    renderedCallback() {
      this.querySelector('x-child').addEventListener('custom', (e) => {
          console.log('Called with ' + JSON.stringify(e.detail));
      });
    }
  }

  /**
   * <template>
   *   <button>Click Me</button>
   * </template>
   */
  export default class ChildElement extends MacroElement {
      renderedCallback() {
          this.querySelector('button').addEventListener('click', this.handleClick);
      }

      handleClick() {

>         this.dispatchEvent(new CustomEvent('custom', { detail: { macro: true } }));

Reviewer 4: given how the event listener is added, what's the expected `this` within the event handler?

Reviewer 3: In this example, the `this` value is the `button` element. Event listeners added via the template are the only ones binding the `this` value to the class.

#### Slots

As mentioned before, the component composition model in LWC is provided by slots. Light DOM will provide the same mental model for developers building Light DOM components.

In Light DOM, `<slot>` will denote the place where the slotted component will be attached. The `<slot>` element itself won't be rendered. The slotted content (or the fallback content) will be flattened to the parent element at runtime.

Since the `<slot>` element itself isn't rendered, adding attributes or event listeners to the `<slot>` element in the template will throw a compiler error.

Due to the above differences from regular `<slot>`, LWC compiler will enforce the presence of `<slot lwc:no-shadow>` on Light DOM slots to make it explicit to the user that these are different slots.

##### Timing

In terms of timing, slots in light DOM will behave similarly to the current synthetic shadow slots. For example:

```html
<!-- x/slottable.html -->
<template lwc:no-shadow>

>   <slot lwc:no-shadow name="foo"></slot>

Reviewer 3: I am not sure if it makes sense to have both the `lwc:no-shadow` attributes on the root template element and on each individual slot. We can either consider that:
- all the slots in the template with the `lwc:no-shadow` attributes are light DOM slots.  
- all the slots are native slots except the ones with the `lwc:no-shadow` attribute.

Reviewer: `<slot>` and `<slot lwc:no-shadow>` cannot exist in the same file, so I think the second option is not viable. Unless `<slot lwc:no-shadow>` were used as a signal that the entire `<template>` is light, which seems odd since it would only apply to slotted components.

That leaves the first option. But my understanding of why we decided to redundantly add `lwc:no-shadow` to the `<slot>` was to match [Aurelia's slots](https://docs.aurelia.io/app-basics/components-revisited#au-slot):

> […] using a different name like `au-slot` makes it clear that the native slot is not used in this case

At this point though, I'm a bit on-the-fence about this, as the redundancy is a bit ugly. Also, Stencil/Vue/Svelte do not seem to think it's necessary to call their slots anything but `<slot>`.

Reviewer: For the record, [Fast](https://www.fast.design/docs/fast-element/working-with-shadow-dom/#:~:text=If%20you%20choose%20to%20render%20to%20the%20Light%20DOM) and Lit disallow `slot`s entirely in light DOM, so we cannot really take inspiration from them.

Reviewer 2: I agree with PM above, but between this thread and [the thread above](https://github.com/salesforce/lwc-rfcs/pull/44/files#r626868899), I'm a little lost where we stand right now if we need the additional directive for slots, or not?

Reviewer: We discussed this and I believe we landed on slots not needing to be marked like this. Any slot in a light-dom template is light; any slot in a shadow-dom template is shadow. I'll update the RFC.

  <slot lwc:no-shadow></slot>
  <slot lwc:no-shadow name="bar"></slot>
</template>
```

```html
<!-- app.html -->
<template>
  <x-slottable>
    <x-baz></x-baz>

>     <x-foo slot="foo"></x-foo>

Reviewer 3: **Side note:**

On the consumer side in the `app.html`, the example below is using the standard `slot` attribute. From the compiler standpoint, it is impossible to know if the `<x-slottable>` uses the light DOM or the shadow DOM slotting algorithm.

I was under the original impression that we could also ask developers to use a custom directive to indicate to the template compiler to use the light DOM slotting algorithm to avoid this branching at runtime. For example using `slot:no-shadow` instead of `slot`. That said I came to the conclusion that it is not possible.

While this custom directive works well with named slots I don't see how this would work with the default slot. In the example below I don't see how it is possible to indicate that the text content should go to the light DOM instead of the shadow DOM.

```html
<template>
  <x-slottable>
    Some content
  </x-slottable>
</template>
```

Reviewer: This is a good point. I don't know how to resolve it, so perhaps slots should be moved to a separate RFC entirely.

Reviewer: After discussing this in a meeting, it seems the only way to resolve the above example is to move the logic to the runtime. I.e. we have to figure out at runtime rather than compile time whether a slot is shadow or light.

The only workaround I can think of would be to make a hard requirement on light DOM slots that they have some top-level element that we can mark:

```html
<template>
  <x-slottable>
    <div lwc:slot> <!-- required -->
      Some content
    </div>
  </x-slottable>
</template>
```

This isn't great for developer ergonomics, but it would allow us to figure this out at compile time.

Reviewer: We decided that whether slots are light or shadow will be determined at runtime.

>     <x-bar slot="bar"></x-bar>

Reviewer 3: Now that I think about this, I don't think the `slot` attribute should be rendered at runtime in the light DOM. 

We decided not to render the `<slot>` element at runtime to prevent potential issues where the light DOM `<slot>`  connects to the parent shadow root. If we were to keep the `slot` attribute at runtime for the light DOM we could also run into unexpected slotting behaviors at runtime when mixing shadow DOM slot and light DOM slots.

Reviewer: Does this run into the same problem where the consumer component needs to know whether `<x-slottable>` is a light DOM element or not? If so, it would maybe argue for `no-shadow:slot="foo"` or `lwc:slot="foo"` or something like that.

Then again as you said above, that does not solve default slots. Unless we force something like `lwc:slot="default"`.

#### Styles

Until now styles used in LWC components were scoped to the component thanks to shadow DOM (or synthetic shadow DOM) style scoping. In the light DOM, component styles naturally leak out of the component; LWC doesn't do any style scoping out of the box. Developers are in charge of writing selectors that are specific enough to target the intended element or pseudo-element.

To support the cases where a shadow DOM element composes a light element, light DOM styles are required to be injected to the closest root node. For a given light DOM element, if all the ancestor components are also light DOM components, the component style sheet will be injected in the document `<head>`. Otherwise, if any of the ancestors is a shadow DOM component, the style has to be injected into the closest shadow root. Upon insertion of a light DOM element, LWC does the following steps:

- look up for the closest root node (`Node.prototype.getRootNode()`)
- insert the stylesheet if not already present:
  - if the root node is the HTML document, the style sheet is inserted as a `<style>` element in the `<head>`.

>   - if the root node is a shadow root, the stylesheet is inserted as a `<style>` element as the first shadow root child.

Reviewer: After trying to implement this, I think this algorithm works for native shadow DOM, but not synthetic shadow DOM. That's because synthetic shadow has no concept of style scoping, so it won't work for the following case:

```html
<x-light-outside>
  <p>I should not be styled</p>
</x-light-outside>
<x-shadow>
  #shadow-root
    <style>p { color: green }</style>
    <x-light-inside>
      <p>I should be styled</p>
    </x-light-inside>
</x-shadow>
```

If `x-light-inside` styles itself, those styles will bleed into `x-light-outside`, but only in synthetic shadow.

To work around this, the algorithm would have to be something like:

1. Look up the closest root node
2. Check if that node is in native shadow or synthetic shadow mode (there could be a mix on the page)
3. If native, just inject a `<style>` tag
4. If synthetic, figure out its hostAttribute/shadowAttribute and generate the scoped styles for that. Then inject those styles into the document `<head>`.

Because of synthetic shadow, it's much trickier than it appears at first glance.

Reviewer: In fact, it's not even clear to me that this is possible to implement in synthetic shadow. Consider:

```html
<x-shadow x-shadow_shadow-host>
    <style>[x-shadow_shadow-host] p { color: green }</style>
    <x-light-inside>
      <p>I should be styled</p>
    </x-light-inside>
    <x-shadow-inside>
      <p>I should not be styled</p>
    </x-shadow-inside>
</x-shadow>
```

In this example, `<x-light-inside>` should have its `<p>` styled, but `<x-shadow-inside>` should not have its `<p>` styled. But I can't think of a CSS rules that would accomplish this, while also correctly bleeding out to other light DOM elements inside the `<x-shadow>`, but not to light DOM components outside `<x-shadow>`.
```

Reviewer: The only alternative that occurs to me is to redefine "global CSS" to mean "entirely global to the whole document." Then we would just always insert bare `<style>`s (no scoping with attributes) into the `<head>`. I'm not sure this is a great long-term plan, though, since in the world of native shadow DOM it's not the ideal solution.

Reviewer: Another problem with synthetic shadow: the root container element is not guaranteed to have styles of its own, meaning it might not have an attribute we can even scope to. We could use the tag name, but then it would apply to all containers with that tag name across the whole DOM, which doesn't work if the light DOM element is only inserted into one of them.

Reviewer: This may be solveable, but only if every light DOM element has attributes added which reflect its containing shadow root component. Working on a fix, but it's starting to look a bit gnarly.

Reviewer: Nevermind, this seems solvable.

Reviewer: Yes, this is resolvable by having the same attribute for all light DOM elements within a given shadow scope.

Reviewer: Just realized this is quite a bit harder at the root level. Consider this:

```html
<body>
  <x-light>
    <div>should be styled</div>
  </x-light>
  <x-shadow> <!-- synthetic shadow -->
    #shadow-root
      <div>should not be styled</div>
  </x-shadow>
  <div>should be styled</div> <!-- not rendered by LWC -->
</body>
```

if `<x-light>` has the CSS `div { color: green }`, then we need to style the two light-DOM `<div>`s but not the shadow one. Assuming the second `<div>` is not rendered by LWC, and the `<x-shadow>` uses synthetic shadow, then I can't think of a CSS selector to scope this properly. Unless we use MutationObserver to listen for all elements added into the root context and apply attributes to them.

Reviewer: Ah wait, just remembered synthetic shadow doesn't protect from light DOM style injection. Nevermind.

Reviewer: We need some kind of clarity on how light DOM styles interact with synthetic shadow, at the global level and within synthetic shadow roots.

Reviewer: I'm also starting to question whether this algorithm is even necessary. In native shadow DOM, we can insert the `<style>` right alongside the element, it works the same as hoisting it into its closest shadow root.

Reviewer: Actually, `:host` is kind of interesting. Imagine this:

```html
<x-shadow>
  <x-light>
    <style>
      :host { background: blue }
    </style>
  </x-light>
</x-shadow>
```

Should `<x-light>` be able to style its shadow container like this? I can't see anything technically _wrong_ with it, although it's a bit weird.

Reviewer: I amended the RFC with more details on how I think global styles should work. I consider this thread resolved.

> It is important to notice that the order in which light DOM components are rendered impacts the order in which stylesheets are injected into the root node and directly influences CSS rule specificity.

Reviewer: This part of the doc needs to be updated with more detail about how this will work for synthetic shadow, and to specify that `:host`/`:host-context()` will not be supported for light DOM.

## Basic example

When the Shadow DOM is turned off for a component, its content is not attached to its shadow-root, but to the element itself. Here is an example, showing whenever Shadow DOM is on or off:

_Shadow DOM_

```html
<app-container-blue-shadow>
  #shadow-root (open)
  | <div>
  |   <b>Blue Shadow:</b>
  |   <span class="counter">...</span>
  |   <button type="button">Add one</button>
  | </div>
</app-counter-blue-shadow>
```

_Light DOM_

```html
<app-container-blue-light>
    <div>
      <b>Blue Light:</b>
      <span class="counter">...</span>
      <button type="button">Add one</button>
    </div>
</app-counter-blue-light>
```

As a result, when the content of a component resides in the Light DOM, it can be accessed like any other content in the Document host, and thus behave like any other content (styling, APIs, accessibility, third party tooling...).

## Motivation

Consumer applications require DOM traversal and observability of an application’s anatomy from the document root. Without this, theming becomes hard and 3rd party applications do not run properly:

> - **Theming and branding**

Reviewer 5: As Dean said in DRB today, UIP has the same (near-term future) requirements around theming and branding for LEX and LWR apps as Communities.

Reviewer 6: I don't believe the LEX requirements are the same as what Communities is requesting because the number of actors in the system is very different. Let's discuss offline.

Reviewer 3: @hsterlingsfdc Could you make sure to bring the LWC team into the loop when this will be discussed/designed? I would like to better understand the LEX and LWR requirements in this regard.

  Because of the CSS isolation, theming is made harder. Typically, theming is done via APIs (CSS properties) and/or CSS theming parts (`::part`). Both come with caveats and issues. Theming is critical for consumer apps, where custom branding is a must have. Some new APIs, like `::theme`, are investigated but they won’t be pervasively available before years.
  [Styling is critical to web component reuse, but may prove difficult in practice](https://component.kitchen/blog/posts/styling-is-critical-to-web-component-reuse-but-may-prove-difficult-in-practice)

- **Third party integrations**

>   Third party tools need to traverse the DOM, which breaks with Shadow DOM and the existing browser APIs (querySelector, ...). Note that the use of Light DOM fixes for Light DOM components, but not for native Shadow DOM ones if the page contains any: Analytics tools, Personalization platforms, Commerce tools like PriceSpider or Honey...

Reviewer: Reworded it to make it clearer. I think it should have been "the use of Light DOM fixes this for Light DOM components," but this should be clear from the rest of the document, so I removed it. a3f6d71

- **Testing software**

  Tools like Selenium, Cypress etc. face the same issues as third party tools when it comes to traversing the DOM.

Introducing components rendering to the Light DOM opens new possibilities where applications can render most of their content in the Light DOM, but also keep individual widgets (e.g. `select` tag) and leaf components that render in the Shadow DOM. This approach combines the best of both worlds by solving the issues presented above and offering strong encapsulation when needed.
![shadow spectrum](./shadow-spectrum.png?raw=true "Shadow Spectrum")

### Prior Art

Most of the libraries designed to support Shadow DOM also propose a Light DOM option, with a variety of Shadow DOM features (slots, scoped styles, etc.). It includes:

- [StencilJS](https://stenciljs.com/docs/styling#shadow-dom-in-stencil)
- [LitElement](https://lit-element.polymer-project.org/api/classes/_lit_element_.litelement.html#createrenderroot)
- [MS Fast Element](https://fast.design/docs/fast-element/working-with-shadow-dom#shadow-dom-configuration)

> ## Detailed design

Reviewer 6: The list of invariants and constraints is missing. From the live discussion on this RFC today it became obvious that there are many you're working with.

* Support all standard Light DOM APIs at runtime
* What the developer codes in the template is what they see in the rendered DOM
* Align with standards where possible; when no standard exist align with pre-existing LWC concepts, and finally  other frameworks 
* A constraint is that the LWC compiler operates on 1 file at a time, not one component (eg HTML + Javascript)

There are many others that need to get captured here. This'll bring visibility to the rationale driving this design.

Reviewer 7: I'm still fuzzy about:

> What the developer codes in the template is what they see in the rendered DOM

what I see in the PR is that the content of the slot is flatten, so the slot element itself never gets rendered.

> ### `LightningElement.shadow`

Reviewer 3: I don't think it's too late to change this, the API is not considered stable today. We debated between @nolanlawson and @abdulsattar this couple of weeks ago. I purposefully made the API name generic because I had in mind to reuse the same `shadow` static field in the future to developers configure the shadow root:

```js
// default - shadow DOM (mode: open)
class extends LightningElement {} 

// shadow DOM (mode: open)
class extends LightningElement {
  static shadow = true;
}

// light DOM
class extends LightningElement {
  static shadow = false;
}

// shadow DOM (mode: closed)
class extends LightningElement {
  static shadow = {
    mode: 'closed'
}
```

@caridy I know that you tinkered with something similar in the past. Where did your previous research lead you?

> ### `LightningElement.shadow`

Reviewer: Another argument in favor of `shadow = false` over `light = true`: web developers tend to think of "shadow DOM" as a fancy new thing, and "light DOM" as the default. In fact, "light DOM" was just "the DOM" for decades.

So developers are more likely to immediately understand `shadow = false` compared to `light = true`, since they understand `shadow` as this weird new bleeding-edge-standards thing that they need to have a separate mental model for, as opposed to `light`, which is just how the web worked for decades.

One counter-argument to that, though, is that developers working with LWC have probably had to acquaint themselves with `shadow` already, as it's the default in LWC. So I see good arguments both ways.

> Toggle between light DOM and shadow DOM is done via a new `shadow` static property on the `LightningElement` constructor. This accepts a `boolean` value and is `true` by default. When set to `true`, the LightningElement creates a shadow root on the host element and renders the component content in the shadow root. When set to `false` the component renders its content directly in the host element light DOM.

Reviewer 8: When dealing with boolean values, as a best practice, the default should be `false`. It looks like this was maybe modeled after Stencil's API which disables shadow DOM by default, whereas we enable it by default. I would prefer to go with something like `static shadowDisabled = true` or `static disableShadow = true`.

Reviewer 3: Interesting. The idea was to make the `shadow` field accept more complex objects in the future. For example, we would also use the `shadow` field to override the shadow DOM mode:

```js
export default class extends LightningElement {
  static shadow = {
    mode: 'closed'
  };
}
```

FYI @caridy as you were working on something similar to configure LWC components.

Reviewer: Agree that defaulting to `true` is weird, but to me the `false` versions feel weirder:

```js
static light = false // "light" what? "light" is an overloaded word
static lightDom = false // verbose, also we now need to choose between lightDom and lightDOM
static noShadow = false // double negative
```

Reviewer 3: Agreed with @nolanlawson on the point above. If we were to change the flag name and use a different mechanism to configure the shadow root, I would vote for @caridy's `static renderToLightDOM: true`.

Reviewer 8: I really only care that the default is `false`, but I'll just point out that the double-negative usually isn't a concern because you don't need to set the property to get the default behavior.

Reviewer 8: After thinking about this some more, I'm ambivalent about `renderToLightDOM`. Although the light DOM doesn't exist in synthetic shadow, both light DOM and shadow DOM do exist in native shadow. We want to indicate that we're going to take what would normally have been rendered in the light DOM and slotted into the shadow DOM, and do some magic without attaching a shadow root. `renderToLightDOM` doesn't seem to capture that, but I don't have any other suggestions at the moment.

Reviewer: This is a good point. `shadow = false` implies "no shadow at all", which is true. Whereas `light = false` implies "no light at all," which is not true because it's actually a mix of light and shadow (due to slots).

```js
import { LightningElement } from "lwc";

// Example of a shadow DOM component
class ShadowDOMComponent extends LightningElement {}

// Example of a light DOM component
class LightDOMComponent extends LightningElement {
  static shadow = false;
}

// Default value
console.log(LightningElement.shadow); // true
```

> The `shadow` property is looked up by the LWC engine when the component is instantiated to determine how it should render. Changing the value of the `shadow` static property after the instantiation doesn't influence whether components are rendered in the light DOM or in the shadow DOM.

Reviewer 6: Does it influence components created after the `shadow` property's value is changed?

Should this value be made non-writable at runtime?

Reviewer: Because of [how we're handling Babel loose mode](https://github.com/salesforce/lwc/blob/f0fb460fd3cacd71e6b702cdb5aa6343491216a1/packages/%40lwc/engine-core/src/framework/base-lightning-element.ts#L603-L615), I'm not sure it's possible to enforce this. I could be wrong though.

As-is though, if `shadow` is redefined after the `constructor()` is called, then the change would have no effect.

Reviewer: Hm, so it looks like it's possible, but we have two options:

1. Disallow setting `shadow` a second time.
2. If any instance of `FooElement` runs its `constructor()`, disallow setting `shadow` again.

The downside of the first is that users may legitimately want to set `shadow` more than once (complex app logic) before calling the constructor. The downside of the second is that it's a bit odd than a single instance of a class would have an impact on the entire class.

Reviewer 3: In the current implementation updating the `shadow` attribute after the component is created has no effect. The attribute is read during the construction phase and then cached in the `VM` for the lifetime of the component.

Making the static field non-writable doesn't provide much guaranty if the field is defined as a getter: 

```js
import { LightningElement } from 'lwc';

export default class FooElement extends LightingElement {
  static get shadow() {
    return Math.random() > 0.5;
  }
}
```

We can for example cache the value `shadow` value after the `FooElement` instantiation and reuse it for all the future `FooElement` instantiations. I am not personally in favor of making the first component instantiation special.

This has some precedent – synthetic shadow always allowed global styles to leak into components (although styles could not leak out of those components). So in a synthetic shadow world, light DOM components' CSS will essentially act like
`<style>`s inserted into the document `<head>`.

#### Scoped styles

Different approaches to layer style scoping on top have been discussed while designing Light DOM, such as introducing a new file extension for automatic style scoping (`.scoped.css`) or using a `<style scoped>` element in the template. This can be addressed in a future RFC.

### Slots

As mentioned before, the component composition model in LWC is provided by slots. Light DOM will provide the same mental model for developers building Light DOM components.

In Light DOM, `<slot>` will denote the place where the slotted component will be attached. The `<slot>` element itself won't be rendered. The slotted content (or the fallback content) will be flattened to the parent element at runtime.

> Since the `<slot>` element itself isn't rendered, adding attributes or event listeners to the `<slot>` element in the template will throw a compiler error.

Reviewer 5: This affects builders if it is not a real slot.  This may mean every single builder (LAB is only one of the many builders Salesforce has) has to add special code to process components dragged into another component's light dom slot.

Reviewer: Unfortunately it would be impossible to use real slots for light DOM, because slots as defined by the web platform are [a part of shadow DOM](https://developer.mozilla.org/en-US/docs/Web/API/Element/slot).

Whatever we decide on for "light DOM slots," it will not be the same as shadow DOM slots.

Reviewer 5: Thanks. Follow-up question: Today, we have internal LWC container components like tabset and accordion that are exposed in LAB (via an aura wrapper).  ISV LWC container components (which are just components that contain slots that the builder understands) are on the roadmap.  It sounds like the recommendation is for ISVs to build these with Light DOM, since they are general purpose and just meant to house other content.

Reviewer: It's really up to each use case. I think we would need to draft a longer document about the strengths and weaknesses of light DOM versus shadow DOM.

For instance, even a tabset component may want to protect its shadow content from being unintentionally styled by the consumer, so shadow DOM may be an appropriate choice there. Also a tabset will inevitably use slots, meaning that there could be performance or timing differences between shadow and light DOM. (E.g. eager versus lazy – a tabset containing 10 slots that only renders 1 might still pay the cost of rendering the other 9, if slots are eager, as they are in native shadow DOM.)

But yes, in general, I think your intuition is correct. Tabsets and accordions don't themselves contain sensitive content, so they could be good candidates for light DOM.

Reviewer 6: I believe the issue is broader than builders: any system that generates LWC components needs to be updated if the LWC syntax differs in how slotted content is provided to a Light DOM-enabled component.

Reviewer 5: Other thoughts:
1. Given the styling/perf differences, we should consider whether some components are configurable by the admin
2. We could use guardrails in the builder to guide admins towards choosing the correct configuration by a) visually showing the difference in preview and b) predicting perf characteristics

Reviewer 2: > any system that generates LWC components needs to be updated

Side note: any system that generates LWC components will need to be updated to respect the `no-shadow` directive anyway.

These are all differences with native shadow slots, which render eagerly. However, these differences align with the current implementation of synthetic shadow slots. This behavior is maintained in light DOM slots for simplicity's sake in the LWC engine, as well as for the performance benefit of lazy slots.

#### Events and other slot features

The `slotchange` event and `::slotted` CSS pseudo-selector are both unsupported for light DOM slots.

In the case of `slotchange`, it's not clear what `event.target` would be. For `::slotted`, it's not really relevant or useful since light DOM slots can be styled the same as any other light DOM content.

### Security

Shadow DOM (in combination with Lightning Locker) encapsulates components and prevents unauthorized access into the shadow tree. With Light DOM though, the DOM is open for traversal by other components. Since Light DOM is not the default, the component author has to opt-in to it, the burden of security falls on them. Developers need to understand that they are "opening" their component for access from outside when they opt in to Light DOM.

> Light DOM will be behind a feature flag that can be set for the runtime. It can be turned on/off on a per container basis. The LWC engine will throw at runtime is a component attempts to render in the light DOM when the feature flag is disabled. 

Reviewer 6: > It can be turned on/off on a per container basis.

Please define container.

> It's important to note that there isn't a way to enable light DOM for a restricted set of namespaces. When the feature flag is turned on, the capability made available to all the components on the page.

Reviewer 6: ```suggestion
It's important to note that there isn't a way to enable light DOM for a restricted set of namespaces. When the feature flag is turned on, the capability is made available to all components on the page.
```

### Server Side Rendering

The engine-server module should provide the SSR capability to seamlessly render Shadow DOM or Light DOM. It should include the component children, as well as generated styles.

## Adoption strategy

This new feature does not break any existing components, it simply adds a new feature that developers have to opt for.

There is no automated migration of the existing components. As discussed above, Light DOM components differs from Shadow DOM components from the API and runtime standpoint. If a component author wishes to use Light DOM for any of their existing components, they will have to make the relevant changes manually.

## How we teach this

Shadow DOM and Light DOM are already names accepted by the industry, see: [Terminology: light DOM vs. shadow DOM](https://developers.google.com/web/fundamentals/web-components/shadowdom?hl=en). We need to provide the proper documentation to educate the LWC developers:

- What are the differences between Shadow DOM and Light DOM

> - We need a guide on when to use one or the other

Reviewer 6: I think flushing this out will clarify the motivation for this capability, which in turn will help reviewers better understand the design and constraints. 

Can you please flush this out in this RFC?

Reviewer: Added some examples of where one might want to use shadow DOM versus light DOM.

> ## Open questions

Reviewer 5: We need to think more about the low-code use cases and the admin persona.

1. How would we protect an admin from using light DOM components that may unknowingly compromise the experience they are building?
2. Does the component schema itself need to indicate whether it is using light DOM?  We have talked about using JSON schema to describe slots, maybe this is part of it
3. How we teach this needs to also address the admin who is building experiences with a mix of light and shadow DOM components.

Reviewer 5: 4. Do we require top-level components exposed in LAB to use shadow DOM?  I don't think we can, supporting LWC container components like tabset seems like a dealbreaker.  Need to think about this more

**How to deal with different namespaced packages claiming the same custom element name?**

As of today on the Salesforce platform, all the LWC components are authored under the `c` namespace. This namespace dictates the prefix for of the custom elements rendered in the DOM. When delivered via a namespaced package, the components are referenced using their actual namespace (eg. `lightning`, `my_ns`, etc.). However internal component references in the package is done using the `c` namespace.

> Because of this if 2 manage package publish a `button` component, the 2 components will be rendered as `c-button` in DOM. This is problematic because a custom element name can only be bound to a single element constructor. Even if this is not a new issue, the [scoped custom element registry proposal](https://github.com/WICG/webcomponents/issues/716) would solve this issue for shadow DOM components. This is still an issue for light DOM component as scoped custom elements registry can only be attached to a shadow root.

Reviewer 6: ```suggestion
Because of this if 2 managed packages publish a `button` component, the 2 components will be rendered as `c-button` in DOM. This is problematic because a custom element name can only be bound to a single element constructor. Even if this is not a new issue, the [scoped custom element registry proposal](https://github.com/WICG/webcomponents/issues/716) would solve this issue for shadow DOM components. This is still an issue for light DOM components because scoped custom elements registry can only be attached to a shadow root.
```

Shadow DOM provides a wonderful, native component encapsulation and composition model.

The two main reasons for using Shadow DOM are:

**Native Component Encapsulation**

The Shadow DOM encapsulation model provides a way for the component author to keep the component internals hidden, with no effect from the broader environment. This means that global styles have no effect on those internals and those internals cannot be queried from outside of the component. This provides a way for widget-type components to be more portable (eg: embedding into 3rd party apps) at the cost of some complexity that makes the creation of a component or an application harder.

**Native Component Composition**

The Shadow DOM gives custom element developers a way to allow consumers to compose some content that it should render. Think about native elements like `<select>`. This is done via `<slot>` elements that are automatically filled by the content from the light DOM. Without Shadow DOM there is no native component composition and this functionality must be provided by the framework.

> > A single component that needs to stand on its own with its own set of functionality is a good candidate for shadow DOM. Salesforce Lightning components are a good example of that. While one or more components as part of an application might not need Shadow DOM, as their intended use is much clearer and their markup less fragile.

Reviewer 2: > While one or more components as part of an application might not need Shadow DOM, as their intended use is much clearer and their markup less fragile.

This sentence is confusing to me.

Author: Removed it.

## Motivation

> Consumer applications require DOM traversal and observability of an application’s anatomy from the document root. Without this, theming becomes hard while 3rd party applications do not run properly:

Reviewer 2: ```suggestion
Consumer applications require DOM traversal and observability of an application’s anatomy from the document root. Without this, theming becomes hard and 3rd party applications do not run properly:
```

![shadow spectrum](./shadow-spectrum.png?raw=true "Shadow Spectrum")

### Prior Art

Most of the libraries designed to support Shadow DOM also propose a Light DOM option, with a variety of Shadow DOM features (slots, scoped styles, etc.). It includes:

- [StencilJS](https://stenciljs.com/docs/styling#shadow-dom-in-stencil)
- [LitElement](https://lit-element.polymer-project.org/api/classes/_lit_element_.litelement.html#createrenderroot)
- [MS Fast Element](https://fast.design/docs/fast-element/working-with-shadow-dom#shadow-dom-configuration)
- ...

Or frameworks that are built for Light DOM offer a Shadow DOM option:

- [Angular](https://angular.io/guide/component-styles#view-encapsulation)
- [React](https://github.com/Wildhoney/ReactShadow) (community library)
- [Vue.js](https://github.com/karol-f/vue-custom-element) (community library)

## Detailed design

### Selecting Light DOM vs Shadow DOM

> The selection of Light DOM vs Shadow DOM is under the control of the component developer. It is done at the component class level, by inheriting a different base class (`MacroElement`, not `LightningElement`). The LWC compiler can also use this information to check any issue and warn the developer.

Reviewer 2: I would clarify that just like in native JS, choosing to append your children to a shadow root or to the element root should be the choice of the author.

Reviewer 2: Maybe with a code example of how this is done in JavaScript.

Author: Added more detail and an example.

  In Light DOM, `<slot>` will denote the place where the slotted component will be attached. The `<slot>` element itself won't be rendered. The slotted content (or the placeholder content) will be flattened to the parent element at runtime.

  Since the `<slot>` element itself isn't rendered, adding any event listeners to the `<slot>` element in the template will throw a compiler error.

- **Scoped Styles**

  Shadow DOM styles are scoped to the component they belong. In Native shadow, it is enforced by the browser while in Synthetic shadow the styles are scoped by some attributes which are also added to the element at runtime.

  In Light DOM, the styles will be scoped not just to the component they belong, but also to its children. This is similar to how Aura styles its components.

  E.g.

  ```javascript
  <x-a>
    <style>
      p { color: blue }
    </style>

    <p item=1></p>

    <x-b>
      <style>
        p { color: red }
      </style>

      <p item=2></p>
      <p item=3></p> <!-- This paragraph is coming from A and slotted into B -->
    </x-b>
  </x-a>
  ```

>   In this case, item 1 will be in blue and item 2 and 3 will be in red. Also note that in `<x-a>`, we can write a selector of higher specificity `x-b p` to override the styles in `x-b`.

Reviewer 2: > Also note that in `<x-a>`, we can write a selector of higher specificity `x-b p` to override the styles in `x-b`.

Is there a simpler way to explain this, something by saying that light DOM components' styles are _merged_? I feel like this part should be crystal clear.

Author: I removed that statement and added more detail in the example itself. Please let me know if it's clearer now.

>   Global scoping is not supported. If desired, a global stylesheet can be injected manually.

Reviewer 2: > Global scoping is not supported.

What is Global Scoping?

Author: Added more detail.

- **`this.template`**

>   In `LightningElement`, `this.template` returns the shadow-root. It will return `null` in `MacroElement`.

Reviewer 2: > It will return `null` in `MacroElement`.

How do I reference the root element in a light dom component?

Reviewer 3: With `LightningElement`, all the properties and methods on the `this` value interacts with the host element (`this.getAttribute`, `this.querySelector`, `this.classList`). And `this.template` allows to interact with the shadow root. However, LWC never exposes the host element itself to the component instance. You can get access to it, via `this.template.host` but it is more a hack than anything else.

The same model is preserved with `MacroElement`. It is possible to interact with the host element using methods and properties on the `this` value. But there is currently no workaround (like `this.template.host`) to access the host element from a `MacroElement` component instance.

If accessing the host element is necessary, I would vote for making it available to both `LitghtningElement` and `MacroElement` in a similar fashion. For example via a `host` getter.

### Security (WIP)

- In some applications, light-dom components may not be allowed... it’s up to the app context
  - **what is the behavior when it’s not allowed?**
  - Some applications might disable light-dom as a whole

>   - Some applications might disable light-dom selectively using a “privileged code” model

Reviewer 2: _way_ more detail necessary here. Does this mean that a component was authored in the LightDom but app requested it in the shadow dom for ex?

### Component Migration

> There is no migration of the existing components needed. The behaviors of existing components using Shadow DOM remain the same.

Reviewer 2: I would tie this into the new `MacroElement` model saying components extending from `LightningElement` have nothing to do 9of that is the case)

> ### Synthetic Shadow DOM

Reviewer 2: I think many people are asking "if everything is light DOM then what purpose does Synthetic Shadow DOM serve? The way I tried to answer that question today to PMs was saying "there will always be components somehow in the shadow DOM so we will always need an emulation." but maybe I am wrong?

Reviewer 3: We are trying to get out of this model, where some components will always need emulation.

Synthetic shadow purpose was created to emulate shadow DOM on browsers that don't support shadow DOM natively. As we are sunsetting IE11 support, we shouldn't need synthetic shadow anymore.

  // will certainly class some unexpected issue.
  static renderMode = `light`;
}
```

### `LightningElement.prototype.template`

When rendering to the shadow DOM, `LightningElement.prototype.template` returns the component associated `ShadowRoot`. When rendering to the light DOM, `LightningElement.prototype.template` value is `null`.

### DOM querying

This proposal doesn't change the way component authors query the light DOM. Component can query their children elements via `LightningElement.prototype.querySelector` and `LightningElement.prototype.querySelectorAll`. It means that turning a shadow DOM component to a light DOM one, all the occurrences of `this.template.querySelector` have to replaced `this.querySelector`.

### Template

Any template associated with a light DOM component should use `lwc:render-mode="light"` to indicate that it is a light DOM template:

```html
<template lwc:render-mode="light">
  <!-- light DOM content -->
</template>
```

Similarly, shadow DOM components can use `lwc:render-mode="shadow"`, but it's not necessary since it's the default value for that attribute.

> Setting the attribute may or may not actually result in compiler differences, but it will be required to allow us the flexibility to have the compiler behave differently for light DOM components versus shadow DOM components.

Reviewer 8: Can we list out some concrete use cases for this?

Author: @ekashida I don't see any reason why anyone would add `lwc:render-mode="shadow"`, since it's the default. But, just for the sake of consistency, we allow it. The value of the directive can either be `shadow` or `light`, nothing else.

Reviewer 8: @abdulsattar I was referring to the need to have a `render-mode` directive. What use case does it solve? Can we mention those use cases here?

Reviewer: Added some text here.

---
title: Light DOM Support
status: DRAFTED
created_at: 2020-09-10
updated_at: 2020-09-10
champion: Philippe Riand (priand), Ted Conn (tconn)
pr: https://github.com/salesforce/lwc-rfcs/pull/44
---

# RFC: Light DOM Support

## Summary

LWC currently enforces the use of Shadow DOM for every component. This proposal aims to provide a new option, a toggle, which instead lets a component attach its content as children in the Light DOM.

> ## Basic example

Reviewer: MacroElement is gone now

- **Theming and branding**

  Because of the CSS isolation, theming is made harder. Typically, theming is done via APIs (CSS properties) and/or CSS theming parts (`::part`). Both come with caveats and issues. Theming is critical for consumer apps, where custom branding is a must have. Some new APIs, like `::theme`, are investigated but they won’t be pervasively available before years.
  [Styling is critical to web component reuse, but may prove difficult in practice](https://component.kitchen/blog/posts/styling-is-critical-to-web-component-reuse-but-may-prove-difficult-in-practice)

- **Third party integrations**

  Third party tools need to traverse the DOM, which breaks with Shadow DOM and the existing browser APIs (querySelector, ...). Note that the use of Light DOM fixes for Light DOM components, but for for native Shadow DOM ones if the page contains some.

  - Analytics tools, Personalization platforms, Commerce tools like PriceSpider or Honey...

- **Testing software**

  They faced the same issues than third party tools when it comes to traverse the DOM

> Furthermore, **the goal of Shadow DOM is not to enclose the entire application in a single shadow-bound element**. We want to build UIs which are comprised of multiple web components, not UIs which are a single, top-level web component. Web components shouldn’t be the fundamental mechanism for building everything in an app, otherwise they negate the usefulness of standard semantic HTML. This is why we think the current model is fundamentally flawed.

Reviewer 3: ```suggestion
Introducing components rendering to the Light DOM opens new possibilities where applications can render most of their content in the Light DOM. But also keep individual widgets and leaf components are rendered in the Shadow DOM. This approach combines the best of both worlds by solving the issues presented above and offering strong encapsulation when needed.
```

> ![shadow spectrum](./shadow-spectrum.png?raw=true "Shadow Spectrum")

Reviewer 3: Image content: Remove "Today" and "Where we want to be" label. The RFC should not present points of view but rather preset facts.

  }
}

// doesn't use Shadow DOM
class MyLightElement extends HTMLElement {
  connectedCallback() {
    this.innerHTML = "<p>My element that does not use Shadow DOM";
  }
}
```

In LWC as well, the developer will retain the choice: they can choose to inherit from `LightningElement` to select Shadow DOM (which already exists) or inherit from a new class, `MacroElement`, to opt into Light DOM.

```js
import { MacroElement } from "lwc";

export default class MyLightComponent extends MacroElement {}
```

`MacroElement` is a new class that will be exported from the `lwc` module. Its API and functionality remain identical to `LightningElement` with the primary difference of not using Shadow DOM.

### Component features when using Light DOM

Some of the LWC component capabilities are directly inherited from Shadow DOM, or emulated by the synthetic-shadow. Despite the use of Light DOM, we’d like to keep these features available, even if their behavior is slightly adapted to the Light DOM:

- **Slots**

  As we mentioned before, the component composition model in LWC is provided by slots. Light DOM will provide the same mental model for developers building Light DOM components.

>   In Light DOM, `<slot>` will denote the place where the slotted component will be attached. The `<slot>` element itself won't be rendered. The slotted content (or the placeholder content) will be flattened to the parent element at runtime.

Reviewer 3: ```suggestion
  In Light DOM, `<slot>` will denote the place where the slotted component will be attached. The `<slot>` element itself won't be rendered. The slotted content (or the fallback content) will be flattened to the parent element at runtime.
```

>   Since the `<slot>` element itself isn't rendered, adding any event listeners to the `<slot>` element in the template will throw a compiler error.

Reviewer 3: ```suggestion
  Since the `<slot>` element itself isn't rendered, adding attributes or event listeners to the `<slot>` element in the template will throw a compiler error.
```

- **Scoped Styles**

>   Shadow DOM styles are scoped to the component they belong. Styles don't leak out to the rest of the page and the page styles don't leak into this component. In Native shadow, it is enforced by the browser whereas in Synthetic shadow it is enforced by adding few attributes to the elements and scoping all CSS rules by those attributes.

Reviewer 3: ```suggestion
In the Shadow DOM, styles are scoped to the enclosing shadow tree. Styles don't leak out to the rest of the page and the page styles don't leak into this component. In Native shadow, it is enforced by the browser whereas in Synthetic shadow it is enforced by adding few attributes to the elements and scoping all CSS rules by those attributes.
```

>   Styles scoping in Light DOM will be done differently. The styles will be scoped not just to the component they belong, but also to its children. These styles won't leak out to the parents (or their children) though. This is similar to how Aura styles its components.

Reviewer 3: ```suggestion
  Styles scoping in Light DOM will be done differently. The styles will be scoped not just to the component they belong, but also to its children. These styles won't leak out to the parents (or their children) though.
```

  E.g.

  ```html
  <x-a>
    <style>
      p {
        color: blue;
      }
    </style>

    <p></p> <!-- will be blue -->
    <p class="red"></p> <!-- will still be blue. red class from child won't leak out here -->

    <x-b>
      <style>
        p.red {
          color: red;
        }
      </style>

      <p></p> <!-- will be blue. Style from x-a will be applied -->
      <p class="red"></p> <!-- will be red -->
      <p class="red"></p> <!-- This paragraph is coming from A and slotted into B. Will also be red. -->
    </x-b>
  </x-a>
  ```

>   Opting out of this scoping is not supported. There's no way for a component author to say a CSS rule shouldn't be scoped to that specific component (and its children). For example, CSS modules provides a `:global` selector that can be used to opt out of this local scoping. LWC won't provide any such selector. If global scoping is desired, a global stylesheet can be injected manually.

Reviewer 3: `:global` is not a standard CSS selector. I don't think it worth referencing this here. 

```suggestion
  Opting out of this scoping is not supported. There's no way for a component author to say a CSS rule shouldn't be scoped to that specific component (and its children). If global scoping is desired, a global stylesheet can be injected manually.
```

- **`this.template`**

  In `LightningElement`, `this.template` returns the shadow-root. It will return `null` in `MacroElement`.

### Security (WIP)

- In some applications, light-dom components may not be allowed... it’s up to the app context
  - **what is the behavior when it’s not allowed?**
  - Some applications might disable light-dom as a whole
  - Some applications might disable light-dom selectively using a “privileged code” model

### Component Migration

There is no migration of the existing components. Existing components inherit from `LightningElement` and light DOM components will inherit from `MacroElement` and they are two different things. If a component author wishes to use Light DOM for any of their existing components, they will have to make the relevant changes manually.

### Synthetic Shadow DOM

The selection of Light DOM should not be impacted by the use of synthetic shadow instead of native shadow. Now, the goal is to get rid of synthetic shadow, but this is hardly possible today because of:

- Salesforce Lightning Components do not work with native shadow today
- Shadow DOM has accessibility issues

> Could we think of a lightweight synthetic shadow that do not override the global methods but provides enough functions to the base components to work while letting third party integration tools work? This can be a different DOM option, which is an hybrid between Shadow DOM and Light DOM.

Reviewer 3: I don't think this section belongs to the proposal. 
It overlaps more with @ravijayaramappa's work on Shadow DOM mixed mode.

> ## Internal Implementation

Reviewer 3: I would extract this section out of the RFC and move this content to a spike document. 
The internal implementation might change over time, while the RFC should remain accurate on the long run.

```html
<template macro>
  <h1>My Light DOM element</h1>

  <slot onslotchange={handleSlotChange}></slot> <!-- throws error -->
  <div lwc:dom="manual"></div> <!-- throws error -->
  </slot>
</template>
```

In the presence of the `macro` attribute, the template-compiler will do additional validations. Currently, there doesn't seem to be a need to modify the compiler output in anyway.

### Scoped Styles

Style scoping will be done at the host level. The host will have a special attribute (similar to the one that is used in Synthetic shadow) and that attribute selector will be used to prefix all the CSS selectors in the style.

`:host` and `:host()` will also be replaced with the host token.

e.g. for the following CSS,

```css
p {
  color: red;
}
```

the output of style-compiler will be (note the addition of `macroSelector`)

```js
function stylesheet(hostSelector, shadowSelector, nativeShadow, macroSelector) {
  return [macroSelector, "p", shadowSelector, " {color: red;}"].join("");
}
export default [stylesheet];
```

which when evaulated as follows:

```js
const macroSelector = isMacroElement ? `[${tokens.hostAttribute}]` : '';

> content = evaluateStylesheetsContent(
>     stylesheets,
>     hostSelector,
>     shadowSelector,
>     !syntheticShadow,
>     macroSelector

Reviewer 3: ```suggestion
content = evaluateStylesheetsContent(
    stylesheets,
    hostSelector,
    shadowSelector,
    !syntheticShadow,
    macroSelector
)
```

It is important to notice that the order in which light DOM components are rendered impacts the order in which stylesheets are injected into the root node and directly influences CSS rule specificity.

##### `:host` and `:host-context()`

> Also worth noting is that `:host` and `host-context()` are inserted as-is. Therefore, a light DOM component can style its

Reviewer 3: I would call out that this is behaving according to the specification. 
The `:host` and `:host-context` also only match if the root node is a shadow root. Those selectors are ignored when injected at the document level.

Reviewer: Good point, I'll say that.

shadow host. For example:

```css
/* x-light.css */
:host { background: blue; }
```

```html
<x-shadow>
  #shadow-root
    <style>:host { background: blue; }</style>
    <x-light></x-light>
</x-shadow>
```

In this case, `<x-light>` is able to style `<x-shadow>` via `:host`. Unlike shadow components, `:host` does not refer to the component but instead to its actual shadow host.

##### Synthetic shadow

> In the case of synthetic shadow, no attempt is made to scope light DOM styles as described above. The styles are simply inserted as-is, and within a synthetic shadow root they will bleed into the rest of the page.

Reviewer 3: > into the rest of the page.

Including shadow component components using synthetic shadow DOM.
